java/com/facebook/soloader/SoLoader.java

```diff
@@ -340,6 +340,10 @@ private static int makePrepareFlags() {
   }
 
   private static synchronized void initSoLoader(@Nullable SoFileLoader soFileLoader) {
+    if (sSoFileLoader != null) {
+      return;
+    }
+
     if (soFileLoader != null) {
       sSoFileLoader = soFileLoader;
       return;
```

Comment:
The new condition combined with the following lines do worry me as there's clearly an intended way of using this function for subsequent initialisations. We won't be able to land it this way, but if you want to add a separate function that performs that check, I'm sure that should be fine.

Reply:
A little background on this diff: Due to SoLoader crashing our app on a fairly significant percentage of devices, we decided to override the default `SoFileLoader` to use `System.loadLibrary()` instead. To my knowledge, this has practically solved all our SoLoader related crashes.

While implementing this workaround, we noticed that `ReactInstanceManagerBuilder.build()` was overwriting our custom loader even though we made sure to set it first. As it turns out, `SoLoader.init()` gets called twice during build:

- `ReactInstanceManagerBuilder.build()`
  - `ReactInstanceManagerBuilder.getDefaultJSExecutorFactory()`
    - `ReactInstanceManager.initializeSoLoaderIfNecessary()`
      - `SoLoader.init()`
  - `ReactInstanceManager.ReactInstanceManager()`
    - `ReactInstanceManager.initializeSoLoaderIfNecessary()`
      - `SoLoader.init()`

Both times, it will create a new `SoFileLoader` instance regardless of whether a previous one already exists. As far as I can see, we cannot simply add a separate function that doesn't do a check. The alternate fix would be to add a parameter for passing in an `SoFileLoader` to `ReactInstanceManagerBuilder` and make sure to pass it all the way down. Honestly, I'm not sure what's the best approach here. Are you absolutely sure you need the current behaviour?

Slightly off topic: Do you know why we have SoLoader in the first place? From what I can tell, it seems to be handling loading of merged libraries, and I understand this is a feature that has been deprecated since Android 5 (Lollipop). Perhaps there should be away to not use SoLoader at all?

Reply:
I've removed the if-check and left `init` public. Please take another look ❤️